File: libraries/__library__/src/main/java/io/github/NadhifRadityo/Library/Utils/CommonUtils.java
package io.github.NadhifRadityo.Library.Utils;

import groovy.lang.Closure;
import groovy.lang.GroovyObject;
import java.io.OutputStream;
import java.net.URI;
import java.net.URL;
import java.text.CharacterIterator;
import java.util.function.Consumer;
import static io.github.NadhifRadityo.Library.LibraryEntry.getContext;

public class CommonUtils {
	protected static GroovyObject __UTILS_IMPORTED__;
	protected static void __UTILS_CONSTRUCT__() {
		if(__UTILS_IMPORTED__ == null)
			__UTILS_IMPORTED__ = ((Closure<GroovyObject>) ((GroovyObject) getContext().getThat()).getProperty("scriptImport")).call("/std$CommonUtils", "std$CommonUtils");
	}
	protected static void __UTILS_DESTRUCT__() {
		if(__UTILS_IMPORTED__ != null)
			__UTILS_IMPORTED__ = null;
	}
	protected static <T> T __UTILS_GET_PROPERTY__(String property) {
		return (T) __UTILS_IMPORTED__.getProperty(property);
	}
	protected static <T> void __UTILS_SET_PROPERTY__(String property, T value) {
		__UTILS_IMPORTED__.setProperty(property, value);
	}
	public static void __INTERNAL_Gradle$Strategies$CommonUtils_purgeThreadLocal(ThreadLocal<?> threadLocal) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_purgeThreadLocal").call(threadLocal);
	}
	public static void purgeThreadLocal(ThreadLocal<?> threadLocal) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("purgeThreadLocal").call(threadLocal);
	}
	public static boolean __INTERNAL_Gradle$Strategies$CommonUtils_equals(Object other) {
		return CommonUtils.<Closure<Boolean>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_equals").call(other);
	}
	public static String __INTERNAL_Gradle$Strategies$CommonUtils_toString() {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_toString").call();
	}
	public static void __INTERNAL_Gradle$Strategies$CommonUtils_construct() {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_construct").call();
	}
	public static void __INTERNAL_Gradle$Strategies$CommonUtils_downloadFile(URL url, OutputStream outputStream) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_downloadFile").call(url, outputStream);
	}
	public static void downloadFile(URL url, OutputStream outputStream) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("downloadFile").call(url, outputStream);
	}
	public static StringBuilder get__INTERNAL_Gradle$Strategies$CommonUtils_downloadProgressCache() {
		return CommonUtils.<StringBuilder>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_downloadProgressCache");
	}
	public static void __INTERNAL_Gradle$Strategies$CommonUtils_destruct() {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_destruct").call();
	}
	public static CharacterIterator get__INTERNAL_Gradle$Strategies$CommonUtils_readableByteCount() {
		return CommonUtils.<CharacterIterator>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_readableByteCount");
	}
	public static Consumer<Long> __INTERNAL_Gradle$Strategies$CommonUtils_newStreamProgress(long totalSize) {
		return CommonUtils.<Closure<Consumer<Long>>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_newStreamProgress").call(totalSize);
	}
	public static Consumer<Long> newStreamProgress(long totalSize) {
		return CommonUtils.<Closure<Consumer<Long>>>__UTILS_GET_PROPERTY__("newStreamProgress").call(totalSize);
	}
	public static URI __INTERNAL_Gradle$Strategies$CommonUtils_urlToUri(URL url) {
		return CommonUtils.<Closure<URI>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_urlToUri").call(url);
	}
	public static URI urlToUri(URL url) {
		return CommonUtils.<Closure<URI>>__UTILS_GET_PROPERTY__("urlToUri").call(url);
	}
	public static int __INTERNAL_Gradle$Strategies$CommonUtils_hashCode() {
		return CommonUtils.<Closure<Integer>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_hashCode").call();
	}
	public static String __INTERNAL_Gradle$Strategies$CommonUtils_humanReadableByteCount(long bytes0) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_humanReadableByteCount").call(bytes0);
	}
	public static String humanReadableByteCount(long bytes0) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("humanReadableByteCount").call(bytes0);
	}
	public static byte[] __INTERNAL_Gradle$Strategies$CommonUtils_hexStringToBytes(String string) {
		return CommonUtils.<Closure<byte[]>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_hexStringToBytes").call(string);
	}
	public static byte[] hexStringToBytes(String string) {
		return CommonUtils.<Closure<byte[]>>__UTILS_GET_PROPERTY__("hexStringToBytes").call(string);
	}
	public static Gradle.GroovyKotlinInteroperability.GroovyKotlinCache<Gradle.Strategies.CommonUtils> get__INTERNAL_Gradle$Strategies$CommonUtils_cache() {
		return CommonUtils.<Gradle.GroovyKotlinInteroperability.GroovyKotlinCache<Gradle.Strategies.CommonUtils>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_cache");
	}
	public static String __INTERNAL_Gradle$Strategies$CommonUtils_bytesToHexString(byte... bytes) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_bytesToHexString").call(bytes);
	}
	public static String bytesToHexString(byte... bytes) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("bytesToHexString").call(bytes);
	}
	public static void __INTERNAL_Gradle$Strategies$CommonUtils_printDownloadProgress(long startTime, long total, long current, long speed) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_printDownloadProgress").call(startTime, total, current, speed);
	}
	public static void printDownloadProgress(long startTime, long total, long current, long speed) {
		CommonUtils.<Closure<Void>>__UTILS_GET_PROPERTY__("printDownloadProgress").call(startTime, total, current, speed);
	}
	public static String __INTERNAL_Gradle$Strategies$CommonUtils_formattedUrl(String url) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("__INTERNAL_Gradle$Strategies$CommonUtils_formattedUrl").call(url);
	}
	public static String formattedUrl(String url) {
		return CommonUtils.<Closure<String>>__UTILS_GET_PROPERTY__("formattedUrl").call(url);
	}
}

File: gradle/src/main/kotlin/Strategies/CommonUtils.kt
package Strategies

import DynamicScripting.Scripting.addInjectScript
import DynamicScripting.Scripting.removeInjectScript
import GroovyKotlinInteroperability.ExportGradle
import GroovyKotlinInteroperability.GroovyInteroperability.prepareGroovyKotlinCache
import GroovyKotlinInteroperability.GroovyKotlinCache
import Strategies.ExceptionUtils.exception
import Strategies.LoggerUtils.ldebug
import Strategies.LoggerUtils.linfo
import Strategies.ProgressUtils.progress
import Strategies.ProgressUtils.progress_id
import Strategies.StreamUtils.copyStream
import java.io.BufferedInputStream
import java.io.OutputStream
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.nio.file.Paths
import java.text.CharacterIterator
import java.text.StringCharacterIterator
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.function.Consumer
import kotlin.math.log10

object CommonUtils {
	@JvmStatic private var cache: GroovyKotlinCache<CommonUtils>? = null
	@JvmStatic private var downloadProgressCache: StringBuilder? = null
	@JvmStatic internal var readableByteCount: CharacterIterator = StringCharacterIterator("kMGTPE")

	@JvmStatic
	fun construct() {
		cache = prepareGroovyKotlinCache(CommonUtils)
		addInjectScript(cache!!)
	}
	@JvmStatic
	fun destruct() {
		removeInjectScript(cache!!)
		cache = null
	}

	@ExportGradle @JvmStatic @Throws(URISyntaxException::class)
	fun urlToUri(url: URL): URI { return URI(url.protocol, url.userInfo, url.host, url.port, url.path, url.query, url.ref) }
	@ExportGradle @JvmStatic @Throws(MalformedURLException::class, URISyntaxException::class)
	fun formattedUrl(url: String): String { return urlToUri(URL(url)).toASCIIString() }

	@ExportGradle
	@JvmStatic
	fun bytesToHexString(bytes: ByteArray): String {
		val builder = StringBuilder()
		for(aByte in bytes) builder.append(((aByte.toInt() and 0xff) + 0x100).toString(16).substring(1))
		return builder.toString()
	}
	@ExportGradle
	@JvmStatic
	fun hexStringToBytes(string: String): ByteArray {
		val bytes = ByteArray(string.length / 2)
		for(i in string.indices step 2)
			bytes[i / 2] = ((string[i].digitToInt(16) shl 4) + string[i + 1].digitToInt(16)).toByte()
		return bytes
	}

	@ExportGradle
	@JvmStatic
	fun purgeThreadLocal(threadLocal: ThreadLocal<*>) {
		for(thread in Thread.getAllStackTraces().keys) { try {
			val FIELD_Thread_threadLocals: Field = thread.javaClass.getDeclaredField("threadLocals")
			FIELD_Thread_threadLocals.isAccessible = true
			val threadLocals = FIELD_Thread_threadLocals.get(thread) ?: continue
			val FIELD_ThreadLocalMap_remove: Method = threadLocals.javaClass.getDeclaredMethod("remove", ThreadLocal::class.java)
			FIELD_ThreadLocalMap_remove.isAccessible = true
			FIELD_ThreadLocalMap_remove.invoke(threadLocals, threadLocal)
		} catch(e: Throwable) { exception(e) } }
	}

	@ExportGradle
	@JvmStatic
	@Throws(Exception::class)
	fun downloadFile(url: URL, outputStream: OutputStream) {
		 progress(progress_id(url, outputStream)).use { prog0 ->
			prog0.inherit()
			prog0.category = CommonUtils::class.java.toString()
			prog0.description = "Downloading file"
			prog0.pstart()
			val source = Paths.get(URI(url.toString()).path).fileName.toString()
			prog0.pdo("Opening connection $source")
			ldebug("Starting to download: $source")
			val connection = url.openConnection()
			val completeFileSize = connection.contentLength.toLong()
			BufferedInputStream(connection.getInputStream()).use { inputStream ->
				prog0.pdo("Downloading $source")
				linfo("Downloading $source... ($url)")
				copyStream(inputStream, outputStream, newStreamProgress(completeFileSize))
			}
		}
	}

	@ExportGradle
	@JvmStatic
	fun newStreamProgress(totalSize: Long): Consumer<Long> {
		return object : Consumer<Long> {
			var startTime = 0L
			var lastTime = 0L
			var lastLength = 0L
			var lastPrint = 0L
			var speed = 0L
			var speeds: LongArray? = null
			var speedsIndex = 0
			var prog0: ProgressUtils.ProgressWrapper? = null
			override fun accept(length: Long) {
				if(length == -1L) {
					startTime = System.currentTimeMillis()
					lastTime = System.currentTimeMillis()
					speeds = LongArray(30)
					Arrays.fill(speeds, -1L)
					prog0 = progress(progress_id(speeds as Any?))
					prog0?.inherit()
					prog0?.category = CommonUtils::class.java.toString()
					prog0?.description = "Stream progress"
					prog0?.pstart()
					return
				}
				if(length == -2L) {
					prog0?.close()
					return
				}
				prog0?.pdo(String.format("%.2f%%", length * 100.0f / totalSize.toFloat()))
				val alpha = 0.9f
				val now = System.currentTimeMillis()
				val deltaTime = Math.max(1, now - lastTime)
				val deltaLength = length - lastLength
				speeds!![speedsIndex++] = deltaLength * 1000 / deltaTime
				if(speedsIndex >= speeds!!.size) speedsIndex = 0
				lastTime = now
				lastLength = length
				if(now - lastPrint < 1000) return
				var result: Long = 0
				val values = speeds
				var counter = 0
				for(value in values!!) {
					if(value == -1L) continue
					counter++
					result += value
				}
				result /= counter.toLong()
				speed = (speed * alpha).toLong() + (result * (1 - alpha)).toLong()
				printDownloadProgress(startTime, totalSize, length, speed)
				lastPrint = now
			}
		}
	}

	@ExportGradle
	@JvmStatic
	fun printDownloadProgress(startTime: Long, total: Long, current: Long, speed: Long) {
		if(downloadProgressCache == null) downloadProgressCache = StringBuilder(200)
		if(total < 0) return
		val eta = if(speed == 0L) Long.MAX_VALUE else (total - current) * 1000 / speed
		val etaHms = if(speed == 0L) "N/A" else String.format(
			"%02d:%02d:%02d", TimeUnit.MILLISECONDS.toHours(eta),
			TimeUnit.MILLISECONDS.toMinutes(eta) % TimeUnit.HOURS.toMinutes(1),
			TimeUnit.MILLISECONDS.toSeconds(eta) % TimeUnit.MINUTES.toSeconds(1)
		)
		val percent = if(total == 0L) 0 else (current * 100 / total).toInt()
		with(downloadProgressCache!!) {
			append(java.lang.String.join("", Collections.nCopies(2 - if(percent == 0) 0 else log10(percent.toDouble()).toInt(), " ")))
			append(String.format(" %d%% [", percent))
			append(java.lang.String.join("", Collections.nCopies(percent, "=")))
			append('>')
			append(java.lang.String.join("", Collections.nCopies(100 - percent, " ")))
			append(']')
			append(java.lang.String.join("", Collections.nCopies(log10(total.toDouble()).toInt() - if(current == 0L) 0 else log10(current.toDouble()).toInt(), " ")))
			append(String.format(" %d/%d, ETA: %s, Speed: %s/s", current, total, etaHms, humanReadableByteCount(speed)))
		}
		linfo(downloadProgressCache.toString().replace("%".toRegex(), "%%"))
		downloadProgressCache!!.setLength(0)
	}

	@ExportGradle
	@JvmStatic
	fun humanReadableByteCount(bytes0: Long): String {
		var bytes = bytes0
		if(-1000 < bytes && bytes < 1000) return "$bytes B"; readableByteCount.index = 0
		while(bytes <= -999950 || bytes >= 999950) { bytes /= 1000; readableByteCount.next() }
		return String.format("%.1f %cB", bytes / 1000.0, readableByteCount.current())
	}
}
